### PR TITLE
Support parquet page filtering for string columns

### PR DESCRIPTION
**Pull request by alamb**
Draft as it builds on tests in https://github.com/apache/arrow-datafusion/pull/4131

# Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/3833

# Rationale for this change

I want to be able to use parquet page index filtering for string datatypes in IOx.

# What changes are included in this PR?

1. Implement page index filtering for string columns
2. Test

# Are there any user-facing changes?

Hopefully faster parquet predicate evaluation on string columns

### CODE REVIEW

File: datafusion/core/src/physical_plan/file_format/parquet/page_filter.rs [hunk 1]
```diff
@@ -421,7 +423,16 @@ macro_rules! get_min_max_values_for_page_index {
                     vec.iter().map(|x| x.$func().cloned()),
                 )))
             }
-            Index::INT96(_) | Index::BYTE_ARRAY(_) | Index::FIXED_LEN_BYTE_ARRAY(_) => {
+            Index::BYTE_ARRAY(index) => {
+                let vec = &index.indexes;
+                let array: StringArray = vec
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/parquet/page_filter.rs:428]: I am not 100% sure if this is ok (like what if the parquet data got mapped to a LargeStringArray? 🤔

File: datafusion/core/tests/parquet_filter_pushdown.rs [hunk 2]
```diff
@@ -266,20 +266,17 @@ async fn single_file_small_data_pages() {
...
+    TestCase::new(&test_parquet_file)
+        .with_name("selective")
+        // predicate is chosen carefully to prune pages 0, 1, 2, 3, 4
+        // pod = 'iadnalqpdzthpifrvewossmpqibgtsuin'
+        .with_filter(col("pod").eq(lit("iadnalqpdzthpifrvewossmpqibgtsuin")))
+        .with_pushdown_expected(PushdownExpected::Some)
+        .with_page_index_filtering_expected(PageIndexFilteringExpected::Some)
+        .with_expected_rows(2574)
+        .run()
+        .await;
```
Comment (alamb) [inline datafusion/core/tests/parquet_filter_pushdown.rs:279]: this now passes!

File: datafusion/core/src/physical_plan/file_format/parquet/page_filter.rs (hunk 1 above)
Comment (liukun4515) [reply datafusion/core/src/physical_plan/file_format/parquet/page_filter.rs:428]: we need check the logical type for the value.
BYTE_ARRAY in the parquet can represent many logical types, such as `DECIMAL`

Comment (liukun4515) [reply datafusion/core/src/physical_plan/file_format/parquet/page_filter.rs:428]: should also support the type `BYTE_ARRAY` in the `null_counts` of `PagesPruningStatistics`

**Comment by alamb** [comment]
Thanks for the comments @liukun4515  and @isidentical  -- I agree about the nullcounts. I will work on this PR more to address your comments

**Comment by liukun4515** [comment]
> 


> Thanks for the comments @liukun4515 and @isidentical -- I agree about the nullcounts. I will work on this PR more to address your comments

maybe you can just implement the case about utf8 type, I can help you to implement other logical data type for example Decimal.

**Comment by Ted-Jiang** [comment]
@alamb  Would your mind i continue on this, with pr on your branch?😄

**Comment by alamb** [comment]
@Ted-Jiang  -- go right ahead (or also feel free to make another PR and I can close this one0  -- I haven't had the time to work on this for the last few days. I will likely get back to it either later this week or next if you don't have a chance to

**Comment by alamb** [comment]
@Ted-Jiang  I am not sure what happened to this branch -- can you please make a new branch /new PR for this feature? I didn't have a chance to work on (or review) your PR today because some other unexpected work appeared.  😢 


There is so much going on in DataFusion I can barely keep up!

**Comment by Ted-Jiang** [comment]
> There is so much going on in DataFusion I can barely keep up!

of course! You've given too much to this community ❤️

**Pull request closed without merging**